src/core/storage/fileio/s3_filesys.cpp

```diff
@@ -0,0 +1,406 @@
+                       << ". read_ptr_: " << read_ptr_ << ". buffer length "
+                       << buffer_.size() << std::endl;
+
+  // setup the variables, lazily
+  if (begin_bytes >= curr_bytes_ &&
+      begin_bytes - curr_bytes_ + read_ptr_ < buffer_.length()) {
+    read_ptr_ = begin_bytes - curr_bytes_;
+    curr_bytes_ = begin_bytes;
+    return;
+  }
+
+  curr_bytes_ = begin_bytes;
+  read_ptr_ = 0;
+  buffer_.clear();
+}
+
+/*
+ * Retry will be handled by AWS SDK.
+ *
+ * If request is successful, return all bytes read or 0 if no bytes left from
+ * s3. If failed, throw turi::io::io_error.
+ *
+ * @param nwant: number of bytes to read.
+ * @return: number of bytes
+ */
+int AWSReadStreamBase::FillBuffer(size_t nwant) {
+  logstream(LOG_INFO) << "FillBuffer: " << nwant << " bytes" << std::endl;
+  auto s3_client = init_aws_sdk_with_turi_env(url_);
+  // start the new buffer
+  read_ptr_ = 0;
+  nwant = std::min(nwant, file_size_ - curr_bytes_);
+  // nothing to read from remote
+  if (nwant == 0) return 0;
+
+  // in case some nuts
+  if (nwant == size_t(-1)) {
+    log_and_throw_io_failure("input size, size_t(-1), is invalid");
+  }
+
+  // Get the object
+  std::stringstream ss;
+  // range is includsive and zero based
+  ss << "bytes=" << curr_bytes_ << '-' << curr_bytes_ + nwant - 1;
+  logstream(LOG_DEBUG) << "GetObject.Range: " << ss.str() << std::endl;
+
+  Aws::S3::Model::GetObjectRequest object_request;
+  object_request.SetRange(ss.str().c_str());
+  object_request.SetBucket(url_.bucket.c_str());
+  object_request.SetKey(url_.object_name.c_str());
+
+  auto get_object_outcome = s3_client.GetObject(object_request);
+  if (get_object_outcome.IsSuccess()) {
+    // Get an Aws::IOStream reference to the retrieved file
+    auto &retrieved_file =
+        get_object_outcome.GetResultWithOwnership().GetBody();
+
+#ifndef NDEBUG
+    retrieved_file.seekg(0, retrieved_file.end);
+    auto retrieved_size = retrieved_file.tellg();
+    retrieved_file.seekg(0, retrieved_file.beg);
+    if (static_cast<size_t>(retrieved_size) != nwant) {
+      log_and_throw_io_failure(ss.str());
+    }
+#endif
+    // Output the first line of the retrieved text file
+    buffer_.resize(nwant);
+    std::memset(const_cast<char *>(buffer_.data()), 0, nwant);
+    // must use read since \0 is used for padding. no
+    // std::istreambuf_iterator<char>
+    retrieved_file.read(const_cast<char *>(buffer_.data()), nwant);
+  } else {
+    auto error = get_object_outcome.GetError();
+    ss.str("");
+    ss << error.GetExceptionName() << ": " << error.GetMessage() << std::endl;
+    logstream(LOG_ERROR) << ss.str() << std::endl;
+    log_and_throw_io_failure(ss.str());
+  }
+
+  return nwant;
+}
+
+void WriteStream::Write(const void *ptr, size_t size) {
+  size_t rlen = buffer_.length();
+  buffer_.resize(rlen + size);
+  std::memcpy(const_cast<char *>(buffer_.data()) + rlen, ptr, size);
+  if (buffer_.length() >= max_buffer_size_) Upload();
```

Comment:
I don't understand what's going on here with the buffer_.  Why only upload then?

Reply:
Ah, you're chunking it so that writing is done with bigger blocks.

Reply:
Yeah, I'm using AWS multipart upload API.